### PR TITLE
[New Scheduler] Add ActivationService

### PR DESCRIPTION
**Pull request by upgle**
## Description

ActivationService is a GRPC service implementation for the scheduler that communicates with the container proxy.

Please refer to this document:
https://cwiki.apache.org/confluence/display/OPENWHISK/ActivationServiceImpl

## Related issue and scope

- [ ] I opened an issue to propose and discuss this change (#????)

## My changes affect the following components

- [x] Scheduler

## Types of changes

- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [ ] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [x] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

File: core/scheduler/src/main/java/Empty.java [hunk 1]
```diff
@@ -0,0 +1,5 @@
+public class Empty {
+    // Workaround for this issue https://github.com/akka/akka-grpc/issues/289
+    // Gradle complains about no java sources.
+    // Note. Openwhisk is using a lower gradle version, so the latest akka-grpc version cannot be used.
```
Comment (bdoyle0182) [inline core/scheduler/src/main/java/Empty.java:4]: Can we fix this? The issue has been fixed

File: core/scheduler/src/main/protobuf/activation.proto [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+option java_multiple_files = true;
+option java_package = "org.apache.openwhisk.grpc";
+option java_outer_classname = "ActivationProto";
+
+package activation;
+//#options
+
+//#services
+service ActivationService {
+
```
Comment (bdoyle0182) [inline core/scheduler/src/main/protobuf/activation.proto:14]: nit: remove extra lines

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/grpc/ActivationServiceImpl.scala [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import org.apache.openwhisk.core.scheduler.queue._
+import org.apache.openwhisk.grpc.{ActivationService, FetchRequest, FetchResponse, RescheduleRequest, RescheduleResponse}
+import spray.json._
+
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContextExecutor, Future}
+import scala.util.Try
+
+class ActivationServiceImpl()(implicit actorSystem: ActorSystem, logging: Logging) extends ActivationService {
+  implicit val requestTimeout: Timeout = Timeout(50.seconds)
```
Comment (bdoyle0182) [inline core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/grpc/ActivationServiceImpl.scala:18]: should this not be hardcoded?

Comment (bdoyle0182) [inline core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/grpc/ActivationServiceImpl.scala:18]: should this not be hardcoded?

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/grpc/ActivationServiceImpl.scala [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+      fqn <- FullyQualifiedEntityName.parse(request.fqn)
+      rev <- DocRevision.parse(request.rev)
+      msg <- ActivationMessage.parse(request.activationMessage)
+    } yield (fqn, rev, msg)).flatMap(Future.fromTry) flatMap { res =>
+      {
+        val key = res._1.toDocId.asDocInfo(res._2)
+        QueuePool.get(MemoryQueueKey(request.invocationNamespace, key)) match {
+          case Some(queueValue) =>
+            // enqueue activation message to reschedule
+            logging.info(
```
Comment (bdoyle0182) [inline core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/grpc/ActivationServiceImpl.scala:33]: Should this be `debug`

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/grpc/ActivationServiceImpl.scala [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+import scala.concurrent.duration._
+import scala.concurrent.{ExecutionContextExecutor, Future}
+import scala.util.Try
+
+class ActivationServiceImpl()(implicit actorSystem: ActorSystem, logging: Logging) extends ActivationService {
+  implicit val requestTimeout: Timeout = Timeout(50.seconds)
+  implicit val ec: ExecutionContextExecutor = actorSystem.dispatcher
+
+  override def rescheduleActivation(request: RescheduleRequest): Future[RescheduleResponse] = {
+    logging.info(this, s"Try to reschedule activation ${request.invocationNamespace} ${request.fqn} ${request.rev}")
```
Comment (bdoyle0182) [inline core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/grpc/ActivationServiceImpl.scala:22]: I think this should be `debug` statement

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/QueueManager.scala [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  override def serialize = ActionMismatch.serdes.write(this).compactPrint
+}
+
+object ActionMismatch extends DefaultJsonProtocol {
+  val asString: String = "action version does not match"
+  def parse(msg: String) = Try(serdes.read(msg.parseJson))
+  implicit val serdes = jsonFormat(ActionMismatch.apply _, "actionMisMatch")
+}
+
+object QueuePool {
```
Comment (bdoyle0182) [inline core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/QueueManager.scala:78]: Is this safe as a global static object being accessed from futures?

File: core/scheduler/src/main/java/Empty.java (hunk 1 above)
Comment (upgle) [reply core/scheduler/src/main/java/Empty.java:4]: I've already tried it, but It requires a higher Gradle version to use the latest akka-grpc.

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/grpc/ActivationServiceImpl.scala (hunk 5 above)
Comment (upgle) [reply core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/grpc/ActivationServiceImpl.scala:22]: Rescheduling is a special case that happens very occasionally. I think we can keep this log with the `info` level.

File: core/scheduler/src/main/protobuf/activation.proto [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    rpc FetchActivation (FetchRequest) returns (FetchResponse) {}
+
+    rpc RescheduleActivation (RescheduleRequest) returns (RescheduleResponse) {}
```
Comment (ningyougang) [inline core/scheduler/src/main/protobuf/activation.proto:17]: For the design document, `RescheduleActivation rpc is for rescheduling when the container proxy cannot process messages`, can give some example here? i think just describe it is enough.

File: core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/QueueManager.scala (hunk 6 above)
Comment (upgle) [reply core/scheduler/src/main/scala/org/apache/openwhisk/core/scheduler/queue/QueueManager.scala:78]: The only field _queuePool(TrieMap[MemoryQueueKey, MemoryQueueValue]) which is included in this object is a thread-safe.

**Comment by style95** [comment]
Seems there are some formatting errors.
```
[/home/travis/build/apache/openwhisk/tools/travis/../../core/scheduler/build/generated/source/proto/main/akkaGrpc/org/apache/openwhisk/grpc/ActivationService.scala]:
       1: file does not include required license header.
       9: line has trailing whitespace.
      10: line has trailing whitespace.
      12: line has trailing whitespace.
      13: line has trailing whitespace.
      15: line has trailing whitespace.
      23: line has trailing whitespace.
      25: line has trailing whitespace.
      27: line has trailing whitespace.
      29: line has trailing whitespace.
      31: line has trailing whitespace.
  [/home/travis/build/apache/openwhisk/tools/travis/../../core/scheduler/build/generated/source/proto/main/akkaGrpc/org/apache/openwhisk/grpc/ActivationServiceClient.scala]:
       1: file does not include required license header.
      37: line has trailing whitespace.
      39: line has trailing whitespace.
      41: line has trailing whitespace.
      43: line has trailing whitespace.
```

**Review by style95 — APPROVED** [review]

**Pull request merged**